engine/src/main/java/com/redhat/cloud/notifications/session/StatelessSessionFactory.java

```diff
@@ -0,0 +1,64 @@
+package com.redhat.cloud.notifications.session;
+
+import org.hibernate.Session;
+import org.hibernate.StatelessSession;
+import org.jboss.logging.Logger;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.persistence.EntityManager;
+import java.util.function.Consumer;
+
+@ApplicationScoped
+public class StatelessSessionFactory {
+
+    private static final Logger LOGGER = Logger.getLogger(StatelessSessionFactory.class);
+
+    @Inject
+    EntityManager entityManager;
+
+    // An Hibernate session must never be shared across multiple threads.
+    private final ThreadLocal<StatelessSession> threadLocalSession = new ThreadLocal<>();
+
+    /**
+     * This method opens a new stateless session or reuses an existing one if available. That session is then passed to
+     * the given consumer which is executed. At the end of the execution, the session is closed.
+     * @param unitOfWork a unit of work
+     */
+    public void withSession(Consumer<StatelessSession> unitOfWork) {
+        try {
+            unitOfWork.accept(getOrCreateSession());
+        } finally {
+            closeSession();
```

Comment:
Wondering what will happen if there are multiple nested calls to `StatelessSessionFactory.withSession`, wouldn't the last one close it all for the others?

i.e.
```java
StatelessSessionFactory.withSession(s1 -> {
  // do stuff with s1
  StatelessSessionFactory.withSession(s2 -> {
    // do stuff with s2
  })
  // do more more stuff with s1 - but it's closed already?
})
```

Reply:
For what I have review so far - It looks like the assumption above is correct as the code calls the session create at the start of the flows.

I'm wondering if we should have a check in `withSession` to verify that there is no session (else above scenario would happen) - if that's the case, is better to fail early (`"hey, don't call this method multiple times, use StatelessSessionFactory.getSession()"`). 

Also, probably should split the `getOrCreateSession` into a private `createSession` and a public `getSession` so that we can check in `getSession` that there is already a session or warn/fail about the invalid usage.

Reply:
> // do more more stuff with s1 - but it's closed already?

Yes it's closed. I'll try to change things a bit and prevent that.